drivers/clock_control/nrf_power_clock.c

```diff
@@ -1,280 +1,327 @@
 
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+static bool is_in_list(sys_slist_t *list, sys_snode_t *node)
+{
+	sys_snode_t *item = sys_slist_peek_head(list);
 
-		if (!(intenset & CLOCK_INTENSET_HFCLKSTARTED_Msk)) {
-			nrf_clock_int_disable(NRF_CLOCK_INT_HF_STARTED_MASK);
+	do {
+		if (item == node) {
+			return true;
+		} else {
+			item = sys_slist_peek_next(item);
 		}
+	} while (item);
 
-		NVIC_ClearPendingIRQ(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
-
-		irq_enable(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
-	} else {
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+	return false;
+}
 
-		nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTART);
-	}
+static int clock_async_start(struct device *dev,
+			     clock_control_subsys_t sub_system,
+			     struct clock_control_async_data *data)
+{
+	const struct nrf_clock_control_config *config =
+						dev->config->config_info;
+	struct nrf_clock_control *clk_data = dev->driver_data;
+	int key;
 
-	/* release resource guard */
-	m16src_grd = 0U;
+	__ASSERT_NO_MSG((data == NULL) ||
+			((data != NULL) && (data->cb != NULL)));
 
-hf_already_started:
-	/* rollover should not happen as start and stop shall be
-	 * called in pairs.
-	 */
-	__ASSERT_NO_MSG(m16src_ref);
+	key = irq_lock();
+	clk_data->ref++;
 
-	stat = NRF_CLOCK_HFCLK_HIGH_ACCURACY | CLOCK_HFCLKSTAT_STATE_Msk;
-	if ((NRF_CLOCK->HFCLKSTAT & stat) == stat) {
-		return 0;
+	if (clk_data->started) {
+		irq_unlock(key);
+		if (data) {
+			data->cb(dev, data->user_data);
+		}
 	} else {
-		return -EINPROGRESS;
-	}
-}
+		if (clk_data->ref == 1) {
+			bool do_start;
 
-static int m16src_stop(struct device *dev, clock_control_subsys_t sub_system)
-{
-	u32_t imask;
+			do_start =  (config->start_handler) ?
+					config->start_handler(dev) : true;
 
-	ARG_UNUSED(sub_system);
+			if (do_start) {
+				nrf_clock_event_clear(config->started_evt);
+				nrf_clock_task_trigger(config->start_tsk);
+				DBG(dev, "Triggered start task");
+			}
+		}
 
-	/* Test for started resource, if so, decrement reference and acquire
-	 * resource guard.
-	 */
-	imask = irq_lock();
+		/* if node is in the list it means that it is scheduled for
+		 * the second time.
+		 */
+		if (data) {
+			if (is_in_list(&clk_data->list, &data->node)) {
+				irq_unlock(key);
+				return -EALREADY;
+			}
 
-	if (!m16src_ref) {
-		irq_unlock(imask);
-		return -EALREADY;
-	}
+			sys_slist_append(&clk_data->list, &data->node);
+		}
 
-	if (--m16src_ref) {
-		irq_unlock(imask);
-		return -EBUSY;
+		irq_unlock(key);
 	}
 
-	if (m16src_grd) {
-		m16src_ref++;
-		irq_unlock(imask);
-		return -EAGAIN;
-	}
 
-	m16src_grd = 1U;
+	return 0;
+}
 
-	irq_unlock(imask);
+static int clock_start(struct device *dev, clock_control_subsys_t sub_system)
+{
+	return clock_async_start(dev, sub_system, NULL);
+}
 
-	/* re-entrancy and mult-context safe, and reference count is zero, */
+static void nrf_power_clock_isr(void *arg);
 
-	nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTOP);
+static int clock_control_init(struct device *hfclk_dev)
+{
```

Comment:
`data->list` should be initialized here (and in `k32_dummy_init`).

Reply:
done